Feature/ac 393 error managment in request

MarioDelgadoM opened the pull request:
## Description ##
add error managment in request for show and control by the state
add viewcomponent for show error and retry

## Type of Change ##
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Unit or UI test

## Modules Branches Image ##
![Captura de Pantalla 2022-10-07 a la(s) 11 24 15](https://user-images.githubusercontent.com/114523650/194576585-344d5a2e-70fa-47ae-abfe-da05dae178d6.png)


## Linked JIRA task or history ##
https://carsales.atlassian.net/browse/AC-391
https://carsales.atlassian.net/browse/AC-393

MarioDelgadoM committed:
add error ,  manage request
add viewcomponent for show error and retry

MarioDelgadoM committed:
add text to string.xml

MarioDelgadoM committed:
add calendar

MarioDelgadoM committed:
fixed to return a date from calendar and make reusable the datepicker

aapavez commented on app/src/main/java/cl/mario/covid/util/Extensions.kt:
```diff
@@ -1,14 +1,27 @@
+    val monthformat = calendar.get(Calendar.MONTH).plus(1).toString().padStart(2,'0')
+    val dayformat = calendar.get(Calendar.DATE).toString().padStart(2,'0')
+    return "${calendar.get(Calendar.YEAR)}-$monthformat-$dayformat"
+}
```
Maybe, it is better to separate in different files. One file for DateExtensions or DateUtils and other about API

aapavez commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -68,20 +70,31 @@
 
             <Button
                 android:id="@+id/button"
                 android:layout_width="wrap_content"
```
This Id button doesn't has a good name , It is not specific.

aapavez commented on app/src/main/java/cl/mario/covid/ui/main/MainActivity.kt:
```diff
@@ -23,29 +19,40 @@ class MainActivity : AppCompatActivity() {
+        }
+    }
+
+    private fun initListener(){
+        binding.apply {
+            button.setOnClickListener {
+                openCalendar()
+            }
+        }
+    }
+
+    private fun openCalendar(){
+        val calendar = CalendarManager(this)
+        calendar.show()
+        calendar.setOnSelectedListener {
+            covidViewModel.getCovidResults(dateToStringApi(it))
+        }
+    }
```
I know you can reduce more this code 😄 
Maybe, you could to make an utils like:
```

private fun initListener() { 
            binding.button.setOnClickListener {
                openCalendar(context) {
                       covidViewModel.getCovidResults(dateToStringApi(it))
              }
            }
        }
```

Why? Because other developer can use this method without to know the steps necessary for completing configuration.

aapavez commented on app/src/main/java/cl/mario/covid/util/CalendarManager.kt:
```diff
@@ -0,0 +1,37 @@
+
+import android.app.DatePickerDialog
+import cl.mario.covid.ui.main.MainActivity
+import java.util.*
+
+class CalendarManager(private val activity: MainActivity) {
+
+    private var dateSelectedListener: ((date: Date) -> Unit)? = null
```
It not recommended to pass an activity reference (It's better to pass a little of information), maybe you could use context  :)

aapavez commented on app/src/main/java/cl/mario/covid/util/CalendarManager.kt:
```diff
@@ -0,0 +1,37 @@
+class CalendarManager(private val activity: MainActivity) {
+
+    private var dateSelectedListener: ((date: Date) -> Unit)? = null
+
+    fun setOnSelectedListener(listener: (date: Date) -> Unit) {
+        dateSelectedListener = listener
+    }
```
Good Idea 🌟

MarioDelgadoM committed:
separate in different files dateutils , apiutils
change the id button for a better one
change the parameter for calendar , send the context instead mainactivity

MarioDelgadoM committed:
Improving the code when calling the DatePickerDialog

MarioDelgadoM committed:
change id retry btn
